query-parser/src/find.rs

```diff
@@ -0,0 +1,134 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+
+///! This module defines the interface and implementation for parsing an EDN
+///! input into a structured Datalog query.
+///!
+///! The query types are defined in the `query` crate, because they
+///! are shared between the parser (EDN -> query), the translator
+///! (query -> SQL), and the executor (query, SQL -> running code).
+///!
+///! The query input can be in two forms: a 'flat' human-oriented
+///! sequence:
+///!
+///! ```clojure
+///! [:find ?y :in $ ?x :where [?x :foaf/knows ?y]]
+///! ```
+///!
+///! or a more programmatically generable map:
+///!
+///! ```clojure
+///! {:find [?y]
+///!  :in [$]
+///!  :where [[?x :foaf/knows ?y]]}
+///! ```
+///!
+///! We parse by expanding the array format into four parts, treating them as the four
+///! parts of the map.
+
+extern crate edn;
+extern crate mentat_query;
+
+use std::collections::BTreeMap;
+
+use self::edn::Value::PlainSymbol;
+use self::mentat_query::{FindSpec, SrcVar, Variable};
+
+use super::error::{FindParseError, FindParseResult};
+use super::util::{values_to_variables, vec_to_keyword_map};
+
+fn parse_find_parts(find: &[edn::Value],
+                    ins: Option<&[edn::Value]>,
+                    with: Option<&[edn::Value]>,
+                    wheres: &[edn::Value]) -> FindParseResult {
+    // :find must be an array of plain var symbols (?foo), pull expressions, and aggregates.
+    // For now we only support variables and the annotations necessary to declare which
+    // flavor of :find we want:
+    //     ?x ?y ?z       = FindRel
+    //     [?x ...]       = FindColl
+    //     ?x .           = FindScalar
+    //     [?x ?y ?z]     = FindTuple
+    //
+    // :in must be an array of sources ($), rules (%), and vars (?). For now we only support the
+    // default source. :in can be omitted, in which case the default is equivalent to `:in $`.
+    // TODO: process `ins`.
+    let source = SrcVar::DefaultSrc;
+
+    // :with is an array of variables. This is simple, so we don't use a parser.
+    let with_vars = with.map(values_to_variables);
+
+    //
+    // :wheres is a whole datastructure.
+
+    Ok(FindSpec::FindRel(vec!()))
+}
+
+fn parse_find_map(map: BTreeMap<edn::Keyword, Vec<edn::Value>>) -> FindParseResult {
+    // Eagerly awaiting `const fn`.
+    let kw_find = edn::Keyword::new("find");
+    let kw_in = edn::Keyword::new("in");
+    let kw_with = edn::Keyword::new("with");
+    let kw_where = edn::Keyword::new("where");
+
+    // Oh, if only we had `guard`.
```

Comment:
You could do different -- not necessarily better -- with `match (map.get(&kw_find), map.get(&kw_where)) ...`.